Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -136,16 +136,49 @@ public final class PlayerInputSystem: System {
       if event.key == .enter {
         if !message.isEmpty {
           try connection?.sendPacket(ChatMessageServerboundPacket(message))
+          guiState.playerMessageHistory.append(message)
+          guiState.currentMessageIndex = nil
         }
         guiState.messageInput = nil
         return true
       } else if event.key == .escape {
         guiState.messageInput = nil
+        guiState.currentMessageIndex = nil
         return true
       } else if event.key == .delete {
         if !message.isEmpty {
           guiState.messageInput?.removeLast()
         }
+      } else if event.key == .upArrow {
+        // If no message is selected, select the above message
+        if let index = guiState.currentMessageIndex, index > 0 {
+          guiState.currentMessageIndex = index - 1
+          guiState.messageInput = guiState.playerMessageHistory[index - 1]
+        } else if guiState.currentMessageIndex == nil {
+          guiState.stashedMessageInput = guiState.messageInput
+          if !guiState.playerMessageHistory.isEmpty {
```

Comment:
If you swap lines 158 and 159 you can actually avoid having this nested if statement. You can put the `!isEmpty` check as an extra condition on line 157. This is possible because stashing the message input only needs to happen if the message actually gets replaced. Avoiding unnecessary double nesting (where it doesn't affect readability) is good because it simplifies the flow of code for someone reading the code (less paths to follow).